### ISSUE TITLE
diagnostic: rustc points to wrong type parameters in error message

### ISSUE THREAD

**Issue by yaahc** [rust-lang/rust#71708]
In some generic code 

https://github.com/yaahc/errtools/commit/bac775562e3450ac085b33dab3b7d35d2b12dc46

I encountered an error that I expected, but a diagnostic that I did not. Specifically I have a trait like this,

```rust
pub trait WrapErr<T, E, E2> {
    // ...

    /// Wrap the error value with a new adhoc error that is evaluated lazily
    /// only once an error does occur.
    fn wrap_err_with<D, F>(self, f: F) -> Result<T, E2>
    where
        D: Display + Send + Sync + 'static,
        E2: From<(E, String)>,
        F: FnOnce() -> D;
}
```

And I was testing this with an error kind pattern where I knew it wouldn't be able to infer the inner error type because only the outer error type is in the signature.

```rust
#[derive(Error, Debug)]
enum PrivateKind {
    #[error("{msg}")]
    Variant1 {
        source: Box<dyn Error + Send + Sync + 'static>,
        msg: String,
    },
}

#[derive(Error, Debug)]
#[error(transparent)]
struct PublicErrorStruct {
    #[from]
    source: PrivateKind,
}

fn do_thing(path: &str) -> Result<String, PublicErrorStruct> {
    let s = std::fs::read_to_string(path)
        .wrap_err_with(|| format!("unable to read file from path: {}", path))?;

    Ok(s)
}
```

However, when `wrap_err_with` actually produced a type annotation needed error it unhelpfully pointed to the generic parameters on the function, rather than the parameters on the trait.

![Screenshot from 2020-04-30 06-10-04](https://user-images.githubusercontent.com/1993852/80714443-cc4e2180-8aa9-11ea-9d49-01f58be09324.png)

If possible I think the compiler should attempt to pinpoint _which_ generic parameter needs a type annotation and make sure it points to the list where that parameter was declared.

cc @estebank

**Comment by estebank** [comment]
I'm not quite sure I can reproduce this anymore. There's a compilation issue, but after applying the following changes, both examples successfully build

```patch
diff --git a/Cargo.toml b/Cargo.toml
index b8d96e2..625f7dd 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,6 +13,6 @@ serde = "1.0.105"
 adhocerr = "0.1.2"
 eyre = "0.3.5"
 serde_json = "1.0.50"
-anyhow = { path = "/home/jlusby/git/rust/anyhow" }
+anyhow = "1.0.29"
 thiserror = "1.0.16"
 displaydoc = "0.1.5"
diff --git a/src/lib.rs b/src/lib.rs
index b7042cf..d4ffc39 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -160,7 +160,7 @@ impl Serialize for SerializeableError<'_> {
         let mut e = serializer.serialize_struct("error", 3)?;
         let msg = self.0.to_string();
         e.serialize_field("msg", &msg)?;
-        e.serialize_field("backtrace", &self.0.backtrace().map(ToString::to_string))?;
+        // e.serialize_field("backtrace", &self.0.backtrace().map(ToString::to_string))?;
         e.serialize_field("source", &self.0.source().map(ErrTools::serialize))?;
         e.end()
     }
@@ -178,7 +178,7 @@ where
         let msg = self.0.to_string();
         e.serialize_field("type", &std::any::type_name::<E>())?;
         e.serialize_field("msg", &msg)?;
-        e.serialize_field("backtrace", &self.0.backtrace().map(ToString::to_string))?;
+        // e.serialize_field("backtrace", &self.0.backtrace().map(ToString::to_string))?;
         e.serialize_field("source", &self.0.source().map(ErrTools::serialize))?;
         e.end()
     }
```

Without the patch, the errors are 

```
error[E0599]: no method named `backtrace` found for reference `&dyn StdError` in the current scope
   --> src/lib.rs:163:48
    |
163 |         e.serialize_field("backtrace", &self.0.backtrace().map(ToString::to_string))?;
    |                                                ^^^^^^^^^ method not found in `&dyn Error`

error[E0599]: no method named `backtrace` found for reference `&E` in the current scope
   --> src/lib.rs:181:48
    |
181 |         e.serialize_field("backtrace", &self.0.backtrace().map(ToString::to_string))?;
    |                                                ^^^^^^^^^ method not found in `&E`
```
